server/src/main/java/org/candlepin/controller/refresher/visitors/ContentNodeVisitor.java

```diff
@@ -87,32 +89,38 @@ public ContentNodeVisitor(ContentCurator contentCurator, OwnerContentCurator own
     public void processNode(NodeProcessor processor, NodeMapper mapper,
         EntityNode<Content, ContentInfo> node) {
 
-        if (node.visited()) {
-            return;
-        }
-
         // Content nodes should not have any children
         if (!node.isLeafNode()) {
             throw new IllegalStateException("Content node has one or more children nodes");
         }
 
+        // If this node already has a state, we don't need to reprocess it (probably)
+        if (node.getNodeState() != null) {
+            return;
+        }
+
         Content existingEntity = node.getExistingEntity();
         ContentInfo importedEntity = node.getImportedEntity();
 
+        // Default the node state to UNCHANGED and let our cases below overwrite this
+        node.setNodeState(NodeState.UNCHANGED);
+
         if (existingEntity != null) {
-            if (importedEntity != null && ContentManager.isChangedBy(existingEntity, importedEntity)) {
-                Content mergedEntity = this.createEntity(mapper, node);
-                node.setMergedEntity(mergedEntity);
-
-                // Store the mapping to be updated later
-                Map<String, String> contentUuidMap = this.ownerContentUuidMap.get(node.getOwner());
-                if (contentUuidMap == null) {
-                    contentUuidMap = new HashMap<>();
-                    this.ownerContentUuidMap.put(node.getOwner(), contentUuidMap);
+            if (importedEntity != null) {
+                if (ContentManager.isChangedBy(existingEntity, importedEntity)) {
+                    Content mergedEntity = this.createEntity(mapper, node);
+                    node.setMergedEntity(mergedEntity);
+
+                    // Store the mapping to be updated later
+                    Map<String, String> contentUuidMap = this.ownerContentUuidMap.get(node.getOwner());
+                    if (contentUuidMap == null) {
```

Comment:
We could you Map.computeIfAbsent here.

Reply:
Ahh, nice. Turns out it's ever-so-slightly faster, too. Fixed here and in a few other places.